### PR TITLE
Remove all usages of Janestreet's Async and Core libraries

### PR DESCRIPTION
**Pull request by psafont**
The only executable user of these was xapi-storage-script, there were libraries using Async, but only to support this user.

The change on this binary is quite invasive, and needed quite a bit of code to support the used interface.

The error handling used in async is translated into Lwt_result type, this is
because the base Lwt only allows exception, which isn't enough to code the
errors produced.

Logging has been changed to use logs with an asynchronous reporter. Now
loglines print the loglevel.

The setup of the loops / promises has been changed, the outer loop has been
removed and its retry-on-error logic has been moved to the two inner watch
loops.

The inotify loop needed quite a few changes as async_inotify was doing complex
handling of events, even consolidating them. This has been simplified by
removing the watches on files and maintaining the directory one, and creating a
list of commands, which the watch loops can act upon, like registering and
deregistering plugins, for example.

The change depends on https://github.com/xapi-project/xs-opam/pull/663 to compile because of a needed fix in inotify

I've tested manually the plugin reload code which registers and unregisters the plugins nicely when they are moved; and a previous version passed the internal GFS2 functional tests. I'll run more tests but I need a formal review to see whether I missed something.

### CODE REVIEW

**Comment by edwintorok** [comment]
Now we finally only have one concurrency library. I don't really mind which of `Lwt` or `Async` we use, as long as we use only one, and not both like we have been doing so far.

File: ocaml/xapi-storage-script/main.ml [hunk 1]
```diff
@@ -11,22 +11,246 @@
...
+      | WEXITED n ->
+          Error (Exit_non_zero n)
+      | WSIGNALED n ->
+          Error (Signal n)
+      | WSTOPPED n ->
+          Error (Signal n)
+  end
+
+  let create ~prog ~args =
+    let args = List.to_seq (prog :: args) |> Array.of_seq in
```
[the remark this answers is not in the document]
Comment (psafont) [reply ocaml/xapi-storage-script/main.ml:165]: Not sure why I did it that way

**Comment by psafont** [comment]
Unfortunately the internal SMAPIv3 tests are failing with this change, it's difficult to triage what's the problem as the binary seems to forward commands as expected. Maybe with tracing it's easy to see what goes wrong, but that's not ready just yet

**Review by lindig — APPROVED** [review]

**Comment by psafont** [comment]
There seems to be a particular problem introduced to pbd.plug, I'll close this PR until I have time to hunt the bug

**Pull request closed without merging**